src/streamlink/plugins/twitch.py

```diff
@@ -462,34 +474,42 @@ def _access_token(self, type="live"):
 
         return sig, token
 
+    def _get_channel_id(self, channel):
+        channel = self.api.user(login=channel)
+        if "_id" in channel:
+            return channel["_id"]
+        raise PluginError("Unable to find channel: {0}".format(channel))
+
     def _check_for_host(self):
-        channel_id = self.api.channel_info(self.channel)["_id"]
+        channel_id = self._get_channel_id(self.channel)
         host_info = self.api.hosted_channel(include_logins=1, host=channel_id).json()["hosts"][0]
-        if "target_login" in host_info:
+        if "target_login" in host_info and host_info["target_login"].lower() != self.channel.lower():
             self.logger.info("{0} is hosting {1}".format(self.channel, host_info["target_login"]))
             return host_info["target_login"]
 
     def _get_hls_streams(self, stream_type="live"):
         self.logger.debug("Getting {} HLS streams for {}".format(stream_type, self.channel))
         self._authenticate()
-        sig, token = self._access_token(stream_type)
+        self._hosted_chain.append(self.channel)
+
         if stream_type == "live":
-            hosted_channel = self._check_for_host()
-            if hosted_channel and self.options.get("disable_hosting"):
+            hosted_channel = not self.options.get("disable_hosting") and self._check_for_host()
+            if self.options.get("disable_hosting"):
                 self.logger.info("hosting was disabled by command line option")
```

Comment:
It's been this way since the `--twitch-disable-hosting` flag was added, if users have been confused they haven't mentioned it (afaik). It's more of a reminder to users that have it set in their `streamlinkrc` file. Having said that, if the log message wasn't there already I wouldn't add it ;) I can't decide if to remove it or not, I like the idea of making less requests, but I'm not sure about removing the existing log message... 

The most optimal way to do it (in the case of no hosting) would be to request the channel stream first, then if it 404s check to see if it's hosting another channel.